gctags/gctags.1
.\" This file is generated automatically by convert.pl from gctags/manual.in.
.TH GCTAGS 18 "April 2002" "GNU Project"
.SH NAME
gctags \- print cross reference list for gtags.
.SH SYNOPSIS
\fBgctags\fP [-bdenrsvw] \fIfile\fP ...
.br
.SH DESCRIPTION
\fBGctags\fP print cross reference list for \fBgtags\fP(1)
from the specified C, C++, YACC, JAVA and assembler source
to standard output.
Each line of output contains the object name,
the line number which it appears, the file in which it is defined,
and a line image separated by white-space.
It's same with the output of \fBctags\fP(1) with \fB-x\fP option.
.PP
Depending upon the options provided to \fBgctags\fP,
objects will consist of function definitions, function references
and other symbols.
.PP
Files whose names end in \'.c\' or \'.h\' are assumed to be C
source files and are searched for C style routine and macro definitions.
Files whose names end in \'.c++\' \'.cc\' \'.cpp\'
\'.cxx\' \'.hxx\' \'.C\' \'.H\' are assumed to be C++
source files.
Files whose names end in \'.y\' are assumed to be YACC source files.
Files whose names end in \'.java\' are assumed to be Java source files.
Files whose names end in \'.s\' or \'.S\' are assumed to be
Assembler source files.
Other files are searched for C style definitions.
.PP
Yacc files each have a special tag. \fByyparse\fP is the start
of the second section of the yacc file.
.PP
This command is the default parser of GLOBAL source code tag system.
.SH OPTIONS
The following options are available:
.TP
\fB-b\fP, \fB--begin-block\fP
Force level 1 block to begin when reach a left brace at the first column.
(C only)
.TP
\fB-d\fP, \fB--define\fP
Pick up not only function but also macro without argument
as a definition.
.TP
\fB-e\fP, \fB--end-block\fP
Force level 1 block to end when reach a right brace at the first column.
(C only)
.TP
\fB-n\fP, \fB--no-tags\fP
Suppress output of tags. It is useful to use with \fB-w\fP option.
.TP
\fB-r\fP, \fB--reference\fP
Locate function references instead of function definitions.
\'GTAGS\' is needed at the current directory.
(C, C++ and Java source only)
By default, locate function definitions.
.TP
\fB-s\fP, \fB--symbol\fP
Collect symbols other than functions.
By default, locate function definitions.
.TP
\fB-v\fP, \fB--verbose\fP
Verbose mode.
.TP
\fB-w\fP, \fB--warning\fP
Print warning message.
.TP
\fB--langmap\fP=map
Language mapping. Each comma-separated  map  consists of
the language name, a colon, and a list of file extensions.
Default mapping is 'c:.c.h,yacc:.y,asm:.s.S,java:.java,cpp:.c++.cc.cpp.cxx.hxx.C.H'.
.PP
The \fB-r\fP and \fB-s\fP options override each other;
the last one specified determines the method used.
.SH DIAGNOSTICS
\fBGctags\fP exits with a non 0 value if an error occurred, 0 otherwise.
Duplicate objects are not considered errors.
.SH "SEE ALSO"
\fBbtreeop\fP(1),
\fBglobal\fP(1),
\fBgtags\fP(1),
\fBhtags\fP(1).
.PP
GNU GLOBAL source code tag system
.br
(http://www.gnu.org/software/global/).
.SH BUG
\fBGctags\fP relies on the input being well formed,
and any syntactical errors will completely confuse it.
.PP
Assembler support is far from complete.  It extracts only ENTRY()
and ALTENTRY() from source file.
Probably valid only for FreeBSD and Linux kernel source.
.SH AUTHOR
Shigio Yamaguchi
.SH HISTORY
The \fBgctags\fP command appeared in FreeBSD 2.2.2.


gctags/gctags.c
/*
 * Copyright (c) 1996, 1997, 1998, 1999
 *            Shigio Yamaguchi. All rights reserved.
 * Copyright (c) 1999
 *            Tama Communications Corporation. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by Tama Communications
 *      Corporation and its contributors.
 * 4. Neither the name of the author nor the names of any co-contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	gctags.c				17-Aug-99
 */

#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

#include "global.h"
#include "gctags.h"

int	bflag;			/* -b: force level 1 block start */
int	dflag;			/* --debug: debug */
int	eflag;			/* -e: force level 1 block end */
int	nflag;			/* -n: doen't print tag */
int	rflag;			/* -r: function reference */
int	sflag;			/* -s: collect symbols */
int	wflag;			/* -w: warning message */
int	yaccfile;		/* yacc file */

const char *progname = "gctags";	/* program name */
char	*notfunction;

int	main(int, char **);
static	void usage(void);

struct words *words;
static int tablesize;

int
main(argc, argv)
	int	argc;
	char	**argv;
{
	char	*p;

	while (--argc > 0 && (++argv)[0][0] == '-') {
		if (!strcmp(argv[0], "--debug")) {
			dflag++;
			continue;
		}
		for (p = argv[0] + 1; *p; p++) {
			switch(*p) {
			case 'b':
				bflag++;
				break;
			case 'e':
				eflag++;
				break;
			case 'n':
				nflag++;
				break;
			case 'r':
				rflag++;
				sflag = 0;
				break;
			case 's':
				sflag++;
				rflag = 0;
				break;
			case 'w':
				wflag++;
				break;
			default:
				usage();
				/* NOTREACHED */
			}
		}
	}
	if (argc < 1)
		usage();
	if (getenv("GTAGSWARNING"))
		wflag++;	
	if (test("r", NOTFUNCTION)) {
		FILE	*ip;
		STRBUF	*sb = stropen();
		int	i;

		if ((ip = fopen(NOTFUNCTION, "r")) == 0)
			die1("'%s' cannot read.", NOTFUNCTION);
		for (tablesize = 0; (p = mgets(ip, NULL, 0)) != NULL; tablesize++)
			strnputs(sb, p, strlen(p) + 1);
		fclose(ip);
		if ((words = malloc(sizeof(struct words) * tablesize)) == NULL)
			die("short of memory.");
		p = strvalue(sb);
		for (i = 0; i < tablesize; i++) {
			words[i].name = p;
			p += strlen(p) + 1;
		}
		qsort(words, tablesize, sizeof(struct words), cmp);
		/* don't call strclose(sb); */
	}
	for (; argc > 0; argv++, argc--) {
		if (!opentoken(argv[0]))
			die1("'%s' cannot open.", argv[0]);
#ifdef _WIN32
		/* Lower case the file name since names are case insensitive */
		strlwr(argv[0]);
#endif
		/*
		 * yacc
		 */
		if (locatestring(argv[0], ".y", MATCH_AT_LAST))
			C(YACC);
		/*
		 * assembler
		 */
		else if (locatestring(argv[0], ".s", MATCH_AT_LAST)	||
			locatestring(argv[0], ".S", MATCH_AT_LAST))
			assembler();
		/*
		 * java
		 */
		else if (locatestring(argv[0], ".java", MATCH_AT_LAST))
			java();
		/*
		 * C++
		 */
		else if (locatestring(argv[0], ".c++", MATCH_AT_LAST)	||
			locatestring(argv[0], ".cc", MATCH_AT_LAST)	||
			locatestring(argv[0], ".cpp", MATCH_AT_LAST)	||
			locatestring(argv[0], ".cxx", MATCH_AT_LAST)	||
			locatestring(argv[0], ".hxx", MATCH_AT_LAST)	||
			locatestring(argv[0], ".C", MATCH_AT_LAST)	||
			locatestring(argv[0], ".H", MATCH_AT_LAST))
			Cpp();
		/*
		 * C
		 */
		else if (locatestring(argv[0], ".c", MATCH_AT_LAST))
			C(0);
		/*
		 * C or C++
		 */
		else if (locatestring(argv[0], ".h", MATCH_AT_LAST)) {
			if (isCpp())
				Cpp();
			else
				C(0);
		}
		closetoken();
	}
	exit(0);
}

static void
usage()
{
	(void)fprintf(stderr, "usage: gctags [-benrsw] file ...\n");
	exit(1);
}

int
cmp(s1, s2)
	const void *s1, *s2;
{
	return strcmp(((struct words *)s1)->name, ((struct words *)s2)->name);
}

int
isnotfunction(name)
	char	*name;
{
	struct words tmp;
	struct words *result;

	if (words == NULL)
		return 0;
	tmp.name = name;
	result = (struct words *)bsearch(&tmp, words, tablesize, sizeof(struct words), cmp);
	return (result != NULL) ? 1 : 0;
}

void
dbg_print(level, s)
	int level;
	const char *s;
{
	if (!dflag)
		return;
	fprintf(stderr, "[%04d]", lineno);
	for (; level > 0; level--)
		fprintf(stderr, "    ");
	fprintf(stderr, "%s\n", s);
}
